docs/html/development/architecture/configuration-files.rst

```diff
@@ -0,0 +1,145 @@
+  - Manipulating a single configuration file
+
+Both of these operations utilize functionality provided the ``Configuration``
+object, which encapsulates all the logic for handling configuration files and
+provides APIs for the same.
+
+The remainder of this section documents the ``Configuration`` class, and
+discusses potential future refactoring ideas.
+
+
+.. _configuration-class:
+
+``Configuration`` class
+=======================
+
+``Configuration`` loads configuration values from sources in the local
+environment: a combination of configuration files and environment variables.
+
+It can be used in two "modes", for reading all the values from the local
+environment and for manipulating a single configuration file. It differentiates
+between these two modes using the ``load_only`` attribute.
+
+The ``isolated`` attribute manipulates which sources are used when loading the
+configuration. If ``isolated`` is ``True``, user-specific configuration files
+and environment variables are not used.
+
+Reading from local environment
+------------------------------
+
+``Configuration`` can be used to read from all configuration sources in the
+local environment and access the values, as per the precedence logic described
+in the :ref:`Config Precedence <config-precedence>` section.
+
+For this use case, the ``Configuration.load_only`` attribute would be ``None``,
+and the methods used would be:
+
+.. py:class:: Configuration
+
+  .. py:method:: load()
+
+    Handles all the interactions with the environment, to load all the
+    configuration data into objects in memory.
+
+  .. py:method:: items()
+
+    Provides key-value pairs (like ``dict.items()``) from the loaded-in-memory
+    information, handling all of the override ordering logic.
+
+At the time of writing, the parts of the codebase that use ``Configuration``
+in this manner are: ``ConfigOptionParser``, to transparently include
+configuration handling as part of the command line processing logic,
+and ``pip config get``, for printing the entire configuration when no
+:ref:`kind <config-kinds>` is specified via the CLI.
+
+Manipulating a single configuration file
+----------------------------------------
+
+``Configuration`` can be used to manipulate a single configuration file,
+such as to add, change or remove certain key-value pairs.
+
+For this use case, the ``load_only`` attribute would be non-None, and would
+represent the :ref:`kind <config-kinds>` of the configuration file to be
+manipulated. In addition to the methods discussed in the previous section,
+the methods used would be:
+
+.. py:class:: Configuration
+
+  .. py:method:: get_value(key)
+
+    Provides the value of the given key from the loaded configuration.
+    The loaded configuration may have ``load_only`` be None or non-None.
+    This uses the same underlying mechanism as ``Configuration.items()`` and
+    does follow the precedence logic described in :ref:`Config Precedence
+    <config-precedence>`.
```

Comment:
I'd put the description of this method above, in the "Reading from local environment" section above, as it works in both cases.